### PR TITLE
boost: Update list of modules

### PR DESCRIPTION
**Pull request by eschnett**
Add various new modules that are available in Boost, most notably "json" which is required for Meson builds.

Make list of default variants more readable, and add "json" to it.

### CODE REVIEW

**Comment by hainest** [comment]
@eschnett I'm working on a PR that will add all of these components and update the existing ones using conditional variants. I just heard back from the Boost folks about a couple of questions I had, so that PR should be ready this week.

As an aside, we don't want to add any more default components in this package. That model has been replaced with a composable one (#28623).

Thanks!

**Comment by eschnett** [comment]
@hainest Do you have a PR?

**Comment by hainest** [comment]
@eschnett Sorry, I have been super busy this week and I still need to talk to Adam about a couple of items before I can move forward.

In the interim, let's merge this with a few modifications

- remove json from `with_default_variants`. The idea of default variants is scheduled to go away soon.
- add back signals to `with_default_variants`. It's super broken, but we'll at least preserve current functionality.
- remove `headers`. It isn't a real library, but a pseudo target for b2. It's defaulted ON in the Jamfile, so we don't need to expose it here. With the new composable model, installing header-only is the default, anyway.

**Comment by eschnett** [comment]
@hainest Done.

File: var/spack/repos/builtin/packages/boost/package.py [hunk 1]
```diff
@@ -91,28 +108,33 @@ class Boost(Package):
...
         'math',
         'mpi',
+        'nowide',
         'program_options',
         'python',
         'random',
         'regex',
         'serialization',
-        'signals',
+        'stacktrace',
```
Comment (hainest) [inline var/spack/repos/builtin/packages/boost/package.py:132]:
```suggestion
        'signals',
        'stacktrace',
```

It shouldn't be there, but signals is still an allowed variant.

**Review by hainest — CHANGES_REQUESTED** [review]

**Review by hainest — APPROVED** [review]

**Comment by hainest** [comment]
@eschnett Thanks!

**Review by alalazo — CHANGES_REQUESTED** [review]
Shouldn't we keep track when a library was added to boost? That is, shouldn't we prohibit things like:
```
boost@1.39.0 +type_erasure
```
(or anything that was added at a later version)?

**Comment by eschnett** [comment]
@alalazo Ideally we should.

This PR adds the `json` package because I have another package (`reprimand`) that requires it (because the `meson` build system requires it), and my primary intent for this PR was to make `reprimand` build again.

**Comment by hainest** [comment]
@alalazo I am working on a PR to convert to conditional variants for everything. This PR is a stopgap so Erik can use one of the missing libraries.

**Review by alalazo — APPROVED** [review]
Thanks for the context, then LGTM

**Pull request merged**